File: 394-decode-string/solution.go
package main

import (
	"strconv"
	"strings"
)

type stack []rune

func (s stack) Top() *rune {
	if len(s) == 0 {
		return nil
	}
	return &s[len(s)-1]
}

func (s stack) Pop() (stack, *rune) {
	if len(s) == 0 {
		return s, nil
	}
	return s[:len(s)-1], &s[len(s)-1]
}

func (s stack) Push(i rune) stack {
	return append(s, i)
}

func IsNum(i rune) bool {
	if int(i) >= 48 && int(i) <= 57 {
		return true
	}
	return false
}

func decodeString(s string) string {
	var stck stack
	for _, i := range s {
		if i != ']' {
			stck = append(stck, i)
		} else {
			var curr []rune
			for len(stck) > 0 && *stck.Top() != '[' {
				curr = append([]rune{*stck.Top()}, curr...)
				stck, _ = stck.Pop()
			}
			stck, _ = stck.Pop()

			var num []rune
			for len(stck) > 0 && IsNum(*stck.Top()) {
				num = append([]rune{*stck.Top()}, num...)
				stck, _ = stck.Pop()
			}

			n, _ := strconv.Atoi(string(num))
			currString := strings.Repeat(string(curr), n)
			stck = append(stck, []rune(currString)...)
		}
	}
	return string(stck)
}

File: 394-decode-string/solution.py
class Solution:
    def getTop(self, stack: list[int]) -> int:
        top = stack[-1] if stack else None
        return top

    def isLetter(self, s:str) -> bool:
        if ord(s) >= 97 and ord(s) <= 122:
            return True
        return False

    def isNum(self, s:str) -> bool:
        if ord(s) >= 48 and ord(s) <= 57:
            return True
        return False

    def decodeString(self, s: str) -> str:
        a = []
        for i in range(len(s)):
            if s[i] != ']':
                a.append(s[i])
            else:
                curr = []
                while a and a[-1] != '[':
                    curr.insert(0, a.pop())
                a.pop() # remove "["
                
                num = []
                while a and self.isNum(a[-1]):
                    num.insert(0, a.pop())

                n = int("".join(num))
                a.extend("".join(curr) * n)
        return "".join(a)